trillian/integration/hammer_test.go

```diff
@@ -0,0 +1,299 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integration
+
+import (
+	"context"
+	"crypto"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/client"
+	"github.com/google/certificate-transparency-go/jsonclient"
+	"github.com/google/certificate-transparency-go/tls"
+	"github.com/google/certificate-transparency-go/trillian/ctfe/configpb"
+	"github.com/google/certificate-transparency-go/x509"
+)
+
+func TestHammer_NotAfter(t *testing.T) {
```

Comment:
This seems to involve an impressive amount of machinery, but I wonder if most of the test benefit could be gotten from something much simpler -- maybe just an internal test of `make {C,Prec}ertChain` & `getNotAfter`?  So no fake servers or actual submissions needed?

Reply:
Originally that's what I had planned to, but as simple as it is, I had a hard time figuring out how to actually do the change correctly (eg, changing the template doesn't work because it gets overridden in subtle ways, its NotAfter is ignored by makePrecertChain, etc). So I found myself actually patching a local hammer and running it against temporal logs, at which point having a test like this seemed a lot better.

Plus, we get some unit test coverage in a bunch of places we didn't have before, and in a way that's a lot more resilient to change than testing internal, private functions.